File: swift/05_LongestPalindromic.swift
// https://leetcode.com/problems/longest-palindromic-substring/
class Solution {
    var start = 0
    var maxLen = 0
    
    func longestPalindrome(_ s: String) -> String {
        if s.count < 2 {
            return s
        }
        
        let arrStr = Array(s)
        for (i, _) in arrStr.enumerated() {
            if (i < arrStr.count - 1) {
                extendPalindromic(arrStr, i, i)
                extendPalindromic(arrStr, i, i + 1)
            }
        }
        return String(arrStr[start..<maxLen+start])
    }
    
    func extendPalindromic(_ s: Array<Character>, _ j: Int, _ k: Int) {
        var x = j
        var y = k
        while (x >= 0 && y < s.count && s[x] == s[y]) {
            x = x - 1
            y = y + 1
        }
        if (maxLen < y - x - 1) {
            start = x + 1
            maxLen = y - x - 1
        }
    }
    
    
}

File: java/05_LongestPalindromic.java
//https://leetcode.com/problems/longest-palindromic-substring/
class Solution {

    private int start, maxLen;

    public String longestPalindrome(String s) {
        int len = s.length();
        if (len < 2) {
            return s;
        }

        for (int i=0; i < len-1; i++) {
            extendPalindrome(s, i, i);
            extendPalindrome(s, i, i + 1);
        }

        return s.substring(start, start + maxLen);
    }

    private void extendPalindrome(String s, int j, int k) {
        while (j >= 0 && k < s.length() && s.charAt(j) == s.charAt(k)) {
            j--;
            k++;
        }

        if (maxLen < k - j - 1) {
            start = j + 1;
            maxLen = k - j - 1;
        }
    }
}